Link headers don't support referrerpolicy

charliewolfe opened the pull request:
#### 03cd94f0c94c76ae5aabc94449b075c8d72afcc1
<pre>
Link headers don&apos;t support referrerpolicy
<a href="https://bugs.webkit.org/show_bug.cgi?id=241059">https://bugs.webkit.org/show_bug.cgi?id=241059</a>
&lt;rdar://94384756&gt;

Reviewed by Darin Adler.

Added support for referrer-policy in link headers.
<a href="https://github.com/whatwg/html/pull/7961">https://github.com/whatwg/html/pull/7961</a>

* Source/WebCore/loader/LinkHeader.cpp:
(WebCore::paramterNameFromString):
(WebCore::LinkHeader::setValue):
* Source/WebCore/loader/LinkHeader.h:
(WebCore::LinkHeader::referrerPolicy const):
* Source/WebCore/loader/LinkLoader.cpp:
(WebCore::LinkLoader::loadLinksFromHeader):
* LayoutTests/imported/w3c/web-platform-tests/preload/preload-referrer-policy-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/preload/preload-referrer-policy.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/preload/resources/echo-referrer.py: Added.
(main):
* LayoutTests/imported/w3c/web-platform-tests/preload/resources/link-header-referrer-policy.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/preload/resources/link-header-referrer-policy.py: Added.
(main):

Canonical link: <a href="https://commits.webkit.org/255354@main">https://commits.webkit.org/255354@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8b8c2d197fbb7480e6dac353874ee7a3c285bd5c

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/92191 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/1420 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/22777 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/101979 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/162220 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/96191 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/1417 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/29815 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/84632 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/98145 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/97847 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/922 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/78714 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/27874 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/82841 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/82499 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/70914 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/36238 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/16464 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/33995 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/17626 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/3711 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/37865 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/40262 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/39764 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/36760 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/loader/LinkLoader.cpp:
```diff
@@ -112,6 +112,10 @@ void LinkLoader::loadLinksFromHeader(const String& headerValue, const URL& baseU
         LinkLoadParameters params { relAttribute, url, header.as(), header.media(), header.mimeType(), header.crossOrigin(), header.imageSrcSet(), header.imageSizes(), header.nonce(), ReferrerPolicy::EmptyString };
+        
+        if (!header.referrerPolicy().isEmpty())
+            params.referrerPolicy = *parseReferrerPolicy(header.referrerPolicy(), ReferrerPolicySource::ReferrerPolicyAttribute);
```
This is not quite correct. We can’t just unconditionally dereference the result of `parseReferrerPolicy`, because that will just crash at runtime. We need to figure out what to do when it does not parse. Could be as simple as using `value_or` to give it a default value.

youennf replied:
Probably we want to keep the default value (empty string in that case).
it seems worth adding a test for bad values.

darinadler reviewed (changes requested):
Looking good, we should finish this.

darinadler commented on Source/WebCore/loader/LinkLoader.cpp:
```diff
@@ -112,6 +112,10 @@ void LinkLoader::loadLinksFromHeader(const String& headerValue, const URL& baseU
 
         LinkLoadParameters params { relAttribute, url, header.as(), header.media(), header.mimeType(), header.crossOrigin(), header.imageSrcSet(), header.imageSizes(), header.nonce(), ReferrerPolicy::EmptyString };
+
+        if (!header.referrerPolicy().isEmpty())
```
Seems like we don’t need this check.

darinadler commented on Source/WebCore/loader/LinkLoader.cpp:
```diff
@@ -112,6 +112,10 @@ void LinkLoader::loadLinksFromHeader(const String& headerValue, const URL& baseU
         LinkLoadParameters params { relAttribute, url, header.as(), header.media(), header.mimeType(), header.crossOrigin(), header.imageSrcSet(), header.imageSizes(), header.nonce(), ReferrerPolicy::EmptyString };
+
+        if (!header.referrerPolicy().isEmpty())
+            params.referrerPolicy = parseReferrerPolicy(header.referrerPolicy(), ReferrerPolicySource::ReferrerPolicyAttribute).value_or(ReferrerPolicy::EmptyString);
```
Can we parse this earlier and and pass it into the `LinkLoadParameters` constructor?

charliewolfe committed:
Link headers don't support referrerpolicy
https://bugs.webkit.org/show_bug.cgi?id=241059
<rdar://94384756>

Reviewed by Darin Adler.

Added support for referrer-policy in link headers.
https://github.com/whatwg/html/pull/7961

* Source/WebCore/loader/LinkHeader.cpp:
(WebCore::paramterNameFromString):
(WebCore::LinkHeader::setValue):
* Source/WebCore/loader/LinkHeader.h:
(WebCore::LinkHeader::referrerPolicy const):
* Source/WebCore/loader/LinkLoader.cpp:
(WebCore::LinkLoader::loadLinksFromHeader):
* LayoutTests/imported/w3c/web-platform-tests/preload/preload-referrer-policy-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/preload/preload-referrer-policy.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/preload/resources/echo-referrer.py: Added.
(main):
* LayoutTests/imported/w3c/web-platform-tests/preload/resources/link-header-referrer-policy.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/preload/resources/link-header-referrer-policy.py: Added.
(main):

Canonical link: https://commits.webkit.org/255354@main
